closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -261,22 +264,47 @@ static void printErrors(void)
     fflush(stderr);
 }
 
+/*
+ * We use a 8 digit map (AABBCCDD) to represent the version of openssl.
+ * AA is the major version,
+ * BB is the minor version,
+ * CC is the revision, and DD for the letter that could be present in any version.
+ * For example, if an openssl version is in this scheme 1.1.1k.
+ * The ossl_ver should be 10101011 ( k is the 11th letter value for DD here... ).
+ */
+static jlong ver_to_l(const char* astring)
+{
+    long major = 0;
+    long minor = 0;
+    long revision = 0;
+    long diff = 0;
+    int num = sscanf(astring, "OpenSSL %ld.%ld.%ld", &major, &minor, &revision);
```

Comment:
I think `last_char` should be acquired here:
```c
    char last_char = 0;
    if (sscanf(astring, "OpenSSL %ld.%ld.%ld%c", &major, &minor, &revision, &last_char) < 3) {
        return -1;
    }
```
The string returned by openssl doesn't have that character at the end, e.g.
```sh
$ openssl version
OpenSSL 1.1.1f  31 Mar 2020
```

Reply:
Added.

Reply:
If we changed to this format, it seems that the version which only contains major and minor will return odd values. Check the following results.
```
1.0 formatted: FFFFFFFF
1.0.0 formatted: 10000000
1.1.1a formatted: 10101001
1.1.1k formatted: 1010100B
1.1.1.z formatted: 10101000
2.0 formatted: FFFFFFFF
5.0 formatted: FFFFFFFF
10.0 formatted: FFFFFFFF
10.1 formatted: FFFFFFFF
```
Hi @keithc-ca , do we need to consider these cases which only contains major and minor?

Reply:
One more thing that raise our attention is `major << 28` for openssl version 16.0:
```
5.0 formatted: 50000000
10.0 formatted: A0000000
10.1 formatted: A0100000
10.0.1a formatted: A0001001
16.0.0 formatted: 0
```
`major << 28` shifts beyond its capacity at 16 for major and becomes 0.